python/mujoco/rollout.cc

```diff
@@ -158,6 +161,47 @@ void _unsafe_rollout(std::vector<const mjModel*>& m, mjData* d, int nroll, int n
   }
 }
 
+// C-style threaded version of _unsafe_rollout
+static ThreadPool* pool = nullptr;
+void _unsafe_rollout_threaded(std::vector<const mjModel*>& m, std::vector<mjData*>& d,
+                              int nroll, int nstep, unsigned int control_spec,
+                              const mjtNum* state0, const mjtNum* warmstart0,
+                              const mjtNum* control, mjtNum* state, mjtNum* sensordata,
+                              int nthread, int chunk_size) {
+  int nfulljobs = nroll / chunk_size;
+  int chunk_remainder = nroll % chunk_size;
+  int njobs = (chunk_remainder > 0) ? nfulljobs + 1 : nfulljobs;
+
+  if (pool == nullptr) {
```

Comment:
Ironically, the management of this pool object is not thread safe.

i.e. if the rollout function is called from multiple threads, and we drop the Python GIL (https://github.com/google-deepmind/mujoco/blob/main/python/mujoco/rollout.cc#L230), then you may end up deleting a thread pool that is being used, or creating two threadpools.

I would lean towards creating a new threadpool on every call to rollout. Have you measured a serious performance issue with that?

Reply:
@nimrod-gileadi, this is clearly stated in the PR description and also discussed in the related design doc 🙂 

The short answer is that yes, there is a measurable cost to starting up the threadpool, especially for smallish tasks. @aftersomemath can perhaps provide some numbers here to guide the discussion. I think the options are:
1. Come up with a mechanism to make this threadsafe, that does not hurt performance or overcomplicate the API.
2. Clearly document that multi-threading is not thread safe.

I think option 2 is fine, even though when pressed I can think of use cases for multi-multi-threaded rollouts. Perhaps @kevinzakka and/or @btaba would have an opinion on options 1 vs 2?

Reply:
I just measured threadpool creation overhead for the case that the pool is recreated every call. I tested the models, hopper, go 2, and humanoid. nroll = [50, 3000, 100] while nstep = [1000, 1, 50] respectively. For hopper the overhead is 2-30%. unitree go 2 is 0.3-4%. humanoid is 0.5-9%. This seems significant, at least in my view.

Yes, the management of the pool is not threadsafe. But also the pool cannot be used from multiple threads because of the way the condition variable and WaitCount work.

Some alternatives that allow a persistent threadpool and allow rollout to be threadsafe are:

1. let the user pass a reference to the pool to rollout (requires wrapping the pool with pybind, or at least managing it)
2. make rollout a function object (easy)
3. let the user start/stop the threadpool and change the threadpools design to allow usage from multiple threads (this will result in a rather complex threadpool implementation)

I'm in favor of implementing both the current approach and, if threading-threaded-rollouts is, important, option 2. The current approach is not mutually exclusive with 2 and 2 is a breaking API change that requires the user to manage an object instead of just calling a function.

Reply:
Technically, it the threadpool is threadsafe, the issue is just that the pool is not compatible with multithreaded usage. My bad, I should have been more clear.

Reply:
I'm ok with just option 2 as well.

Reply:
Pardon my ignorance, can someone spell out Option 2 more clearly?

Reply:
Usage would look like this if implemented exactly as discussed above:

```python
rollout = rollout.Rollout() # creates an object Rollout that acts like a function
rollout(models, data) # Creates or reuses member variable `rollout.pool`
rollout = None # reference lost, the object will be destroyed and the thread pool shut down cleanly
```

Usage could also look like this:
```python
with Rollout() as rollout:
    rollout(models, data)
```

In either case the thread pool will be shut down properly and automatically. Currently, this does not happen and the OS has to shut down the threads when the interpreter exits.

Also threading-threaded-rollout would be possible by using multiple Rollout instances.

At first glance, `rollout.rollout()` would still exist, but it would have to start and stop its thread pool every call. The result is that otherwise functional algorithms would not be able to use persistent threadpools without an additional argument. This seems verbose.

A solution (which I am inclined towards) is to have `rollout.rollout()` create both a persistent Rollout instance and a cleanup handler for it (with [`atexit`](https://docs.python.org/3/library/atexit.html)) that runs on interpreter shutdown.

Reply:
I get now, thank you. 

I'll let @nimrod-gileadi comment on the details but I will say this: the current single-threadedness and pain of using Python threads mean that this module is not highly used in the wild and when it is, it's by expert users. Ergo, it's okay to break existing behaviour. As long as the changelog includes a "breaking changes" section with migration instructions for existing code, we're good. No need to bend over backwards so existing code runs unmodified.

Reply:
I only saw this discussion after I posted my comment. 

I think making the `rollout` an object is a good solution as it also gives me the option to start and stop the threadpool whenever I like.

Reply:
The examples you posted look good, @aftersomemath. Thanks. It would be good to have some configuration options in the Rollout constructor, to choose threadpool size, whether to use a threadpool at all, etc.

@yuvaltassa, I think it's very easy to maintain the current behaviour of the rollout function, as @aftersomemath said. I wouldn't say it's "bending over backwards" and I don't see a reason to break it unnecessarily.